### PR TITLE
[SPARK-10292][SQL] make metadata query-able by adding meta table

### PR DESCRIPTION
**Pull request by cloud-fan**
To begin with, we can set up 2 meta tables: `catalog_tables` and `catalog_columns`.
`catalog_tables` have 3 columns: `name`, `database`, `is_temporary`.
`catalog_columns` have 5 columns: `name`, `table`, `database`, `data_type`, `nullable`.

We can add more meta tables in the future.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 1]
```diff
@@ -44,7 +90,20 @@ trait Catalog {
...
 
-  def lookupRelation(
+  final def lookupRelation(
+      tableIdentifier: Seq[String],
+      alias: Option[String] = None): LogicalPlan = {
+    val tableIdent = processTableIdentifier(tableIdentifier)
+    if (tableIdent.length == 1 && tableIdent.head == TABLES_META_NAME) {
+      LocalRelation(tablesMetaOutput, tablesMeta().map(tablesMetaConverter))
+    } else if (tableIdent.length == 1 && tableIdent.head == COLUMNS_META_NAME) {
+      LocalRelation(columnsMetaOutput, columnsMeta().map(columnsMetaConverter))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: Should we implement it with `RunnableCommand`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 2]
```diff
@@ -44,7 +90,20 @@ trait Catalog {
...
+    } else {
+      lookupRelationInternal(tableIdentifier, alias)
+    }
+  }
+
+  protected def lookupRelationInternal(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: need some refactor here to git rid of this method, I'll try stackable trait later.

**Comment by cloud-fan** [comment]
retest this please.

**Comment by rxin** [comment]
What do you have in mind "add dataframe version"? Maybe it is ok to just ask users to do table("...").

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please.

**Comment by yhuai** [comment]
@cloud-fan Close this one for now?

**Comment by cloud-fan** [comment]
will revist it when I have time...

**Pull request closed without merging**